File: stratum-mcp/src/stratum_mcp/guard/store.py
"""Guard persistence + concurrency + tamper-evident ledger (S1).

Layout (mirrors ``~/.stratum/flows/``)::

    ~/.stratum/guards/<hash(resource_id)>/registry.json   # checksummed policy + initial + cache of current_state
    ~/.stratum/guards/<hash(resource_id)>/ledger.jsonl     # hash-chained, append-only, fsync'd
    ~/.stratum/guards/<hash(resource_id)>/.lock            # flock sidecar (cross-process)

Key correctness properties (see blueprint Trust boundary + S1):
  * The *ledger* is the source of truth for ``current_state`` — ``registry.json``'s
    copy is a best-effort cache. The durable, fsync'd ledger append is the commit point.
  * ``read_ledger`` tolerates a torn trailing line (crash mid-append → recover to last
    durable entry) but treats an interior chain break as tampering (``LedgerCorrupt``).
  * Per-resource serialization uses BOTH an in-process ``asyncio.Lock`` (coroutine
    safety, like ``parallel_exec``) AND a cross-process ``fcntl.flock`` (separate OS
    processes). Blocking ``flock`` is acquired via ``asyncio.to_thread`` so the event
    loop is never blocked.
  * The resource dir is named by ``sha256(resource_id)`` (collision-proof vs slugs);
    the raw id is stored and verified on load.
"""

from __future__ import annotations

import asyncio
import dataclasses
import fcntl
import hashlib
import json
import os
import re
import tempfile
from contextlib import asynccontextmanager
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Optional

from ..result_cache import canonical_json
from .errors import LedgerCorrupt, ResourceIdMismatch

# Module global so tests can ``monkeypatch.setattr(store, "GUARDS_DIR", tmp_path)``
# (mirrors executor._FLOWS_DIR). Read at call time via _guards_root().
GUARDS_DIR: Path = Path.home() / ".stratum" / "guards"

_STATE_NAME_RE = re.compile(r"^[A-Za-z0-9_.-]+$")

# In-process per-resource locks. Keyed by raw resource_id. Mirrors the per-flow
# asyncio.Lock pattern in parallel_exec.py.
_locks: dict[str, asyncio.Lock] = {}


def _guards_root() -> Path:
    return GUARDS_DIR


def _resource_hash(resource_id: str) -> str:
    return hashlib.sha256(resource_id.encode("utf-8")).hexdigest()[:32]


def _validate_resource_id(resource_id: str) -> None:
    if not isinstance(resource_id, str) or not resource_id:
        raise ValueError("resource_id must be a non-empty string")
    if "\x00" in resource_id:
        raise ValueError("resource_id must not contain NUL")
    if resource_id in (".", ".."):
        raise ValueError("resource_id cannot be '.' or '..'")


def resource_dir(resource_id: str) -> Path:
    """Collision-proof directory for a resource. Raw id stored in registry.json
    and verified on load, so a (vanishingly unlikely) hash prefix collision can
    never silently share state — it surfaces as ResourceIdMismatch."""
    _validate_resource_id(resource_id)
    return _guards_root() / _resource_hash(resource_id)


def is_valid_state_name(name: str) -> bool:
    return isinstance(name, str) and bool(_STATE_NAME_RE.match(name))


# --------------------------------------------------------------------------- #
# Dataclasses
# --------------------------------------------------------------------------- #


@dataclass
class GuardRegistry:
    """The registered, checksummed policy for one guarded resource."""

    resource_id: str
    graph: dict[str, list[str]]
    edge_predicates: dict[str, list[dict[str, Any]]]
    initial: str
    terminal: list[str] = field(default_factory=list)
    stakes: dict[str, str] = field(default_factory=dict)
    checksum: str = ""
    graph_version: int = 1
    workspace_root: Optional[str] = None
    # Non-authoritative cache; ledger head is the source of truth.
    current_state: str = ""

    def to_dict(self) -> dict[str, Any]:
        return dataclasses.asdict(self)

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "GuardRegistry":
        known = {f.name for f in dataclasses.fields(cls)}
        return cls(**{k: v for k, v in d.items() if k in known})


@dataclass
class LedgerEntry:
    """One append-only transition/deviation record. ``entry_digest`` is the
    receipt token (returned to callers as ``ledger_ref``)."""

    ts_ms: int
    from_state: str
    to_state: str
    outcome: str  # "applied" | "refused" | "deviation" | "review_clean" | "graph_version"
    kind: str  # "transition" | "deviation" | "graph_version"
    resolved_by: str = "agent"
    idempotency_key: Optional[str] = None
    payload_digest: Optional[str] = None
    rationale: Optional[str] = None
    # The full JudgeResult-shaped verdict for this attempt — stored so an
    # idempotent replay returns the ORIGINAL decision, not a synthesized one.
    verdict: Optional[dict] = None
    prev_digest: str = ""
    entry_digest: str = ""

    def core(self) -> dict[str, Any]:
        """The digested payload — every field EXCEPT entry_digest."""
        d = dataclasses.asdict(self)
        d.pop("entry_digest", None)
        return d

    def to_dict(self) -> dict[str, Any]:
        return dataclasses.asdict(self)

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "LedgerEntry":
        known = {f.name for f in dataclasses.fields(cls)}
        return cls(**{k: v for k, v in d.items() if k in known})


def compute_entry_digest(entry_core: dict[str, Any], prev_digest: str) -> str:
    canonical = canonical_json(entry_core)
    if canonical is None:
        raise ValueError("ledger entry is not JSON-serializable")
    return hashlib.sha256((canonical + prev_digest).encode("utf-8")).hexdigest()


# --------------------------------------------------------------------------- #
# Atomic write (self-contained copy of migrate._atomic_write)
# --------------------------------------------------------------------------- #


def _atomic_write(path: Path, content: str) -> None:
    """Write content to path atomically via tempfile + fsync + os.replace()."""
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = tempfile.NamedTemporaryFile(
        mode="w", dir=path.parent, delete=False, suffix=".tmp", encoding="utf-8"
    )
    try:
        tmp.write(content)
        tmp.flush()
        os.fsync(tmp.fileno())
        tmp.close()
        os.replace(tmp.name, path)
    except Exception:
        try:
            os.unlink(tmp.name)
        except OSError:
            pass
        raise


# --------------------------------------------------------------------------- #
# Registry IO
# --------------------------------------------------------------------------- #


def registry_exists(resource_id: str) -> bool:
    return (resource_dir(resource_id) / "registry.json").exists()


def persist_registry(reg: GuardRegistry) -> None:
    path = resource_dir(reg.resource_id) / "registry.json"
    _atomic_write(path, json.dumps(reg.to_dict(), indent=2, sort_keys=True))


def _load_registry_raw(resource_id: str) -> Optional[GuardRegistry]:
    path = resource_dir(resource_id) / "registry.json"
    if not path.exists():
        return None
    try:
        payload = json.loads(path.read_text())
    except (json.JSONDecodeError, OSError):
        return None
    reg = GuardRegistry.from_dict(payload)
    if reg.resource_id != resource_id:
        raise ResourceIdMismatch(
            f"registry stores resource_id {reg.resource_id!r} but {resource_id!r} requested"
        )
    return reg


def load_registry(resource_id: str) -> Optional[GuardRegistry]:
    """Load policy from registry.json and set ``current_state`` from the ledger
    head (source of truth). Raises LedgerCorrupt on interior chain break."""
    reg = _load_registry_raw(resource_id)
    if reg is None:
        return None
    entries = read_ledger(resource_id)
    reg.current_state = current_state_from_ledger(entries, reg.initial)
    return reg


# --------------------------------------------------------------------------- #
# Ledger
# --------------------------------------------------------------------------- #


def _ledger_path(resource_id: str) -> Path:
    return resource_dir(resource_id) / "ledger.jsonl"


def read_ledger(resource_id: str) -> list[LedgerEntry]:
    """Read + chain-verify the ledger.

    Torn-tail recovery: if the FINAL line fails to parse or breaks the chain (an
    interrupted append), it is dropped and we recover to the last durable entry.
    A chain break at any NON-final line is genuine tampering → LedgerCorrupt.
    """
    path = _ledger_path(resource_id)
    if not path.exists():
        return []
    raw_lines = [ln for ln in path.read_text().splitlines() if ln.strip()]
    entries: list[LedgerEntry] = []
    prev = ""
    for idx, line in enumerate(raw_lines):
        is_last = idx == len(raw_lines) - 1
        try:
            obj = json.loads(line)
            entry = LedgerEntry.from_dict(obj)
        except (json.JSONDecodeError, TypeError, KeyError):
            if is_last:
                break  # torn trailing line — recover
            raise LedgerCorrupt(f"ledger line {idx} is not valid JSON (interior)")
        expected = compute_entry_digest(entry.core(), prev)
        if entry.entry_digest != expected or entry.prev_digest != prev:
            if is_last:
                break  # interrupted append of the final entry — recover
            raise LedgerCorrupt(
                f"ledger chain broken at line {idx} (interior tampering)"
            )
        entries.append(entry)
        prev = entry.entry_digest
    return entries


def verify_chain(entries: list[LedgerEntry]) -> bool:
    prev = ""
    for entry in entries:
        if entry.prev_digest != prev:
            return False
        if entry.entry_digest != compute_entry_digest(entry.core(), prev):
            return False
        prev = entry.entry_digest
    return True


def append_ledger(resource_id: str, entry: LedgerEntry) -> str:
    """Durably append a hash-chained entry. Caller MUST hold the resource lock.

    Returns the entry_digest (the receipt token / ledger_ref).
    """
    entries = read_ledger(resource_id)
    prev = entries[-1].entry_digest if entries else ""
    entry.prev_digest = prev
    entry.entry_digest = compute_entry_digest(entry.core(), prev)

    path = _ledger_path(resource_id)
    path.parent.mkdir(parents=True, exist_ok=True)
    line = canonical_json(entry.to_dict())
    if line is None:
        raise ValueError("ledger entry is not JSON-serializable")
    fd = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_APPEND, 0o644)
    try:
        os.write(fd, (line + "\n").encode("utf-8"))
        os.fsync(fd)
    finally:
        os.close(fd)
    return entry.entry_digest


def find_by_idempotency_key(
    resource_id: str, key: Optional[str]
) -> Optional[LedgerEntry]:
    if not key:
        return None
    for entry in reversed(read_ledger(resource_id)):
        if entry.idempotency_key == key:
            return entry
    return None


def current_state_from_ledger(entries: list[LedgerEntry], initial: str) -> str:
    """The current state is the to_state of the last entry that moved state."""
    for entry in reversed(entries):
        if entry.outcome in ("applied", "deviation"):
            return entry.to_state
    return initial


# --------------------------------------------------------------------------- #
# Locking (in-process asyncio.Lock + cross-process fcntl.flock)
# --------------------------------------------------------------------------- #


def _get_inproc_lock(resource_id: str) -> asyncio.Lock:
    lock = _locks.get(resource_id)
    if lock is None:
        lock = asyncio.Lock()
        _locks[resource_id] = lock
    return lock


@asynccontextmanager
async def resource_lock(resource_id: str):
    """Serialize all access to one resource across coroutines AND OS processes.

    The blocking ``flock`` is acquired on a worker thread so the event loop keeps
    running. The in-process ``asyncio.Lock`` prevents two coroutines in THIS
    process from both holding distinct fds that flock would (on some platforms)
    consider compatible, and gives fair coroutine ordering.
    """
    inproc = _get_inproc_lock(resource_id)
    await inproc.acquire()
    rdir = resource_dir(resource_id)
    rdir.mkdir(parents=True, exist_ok=True)
    lock_path = rdir / ".lock"
    fd = os.open(lock_path, os.O_WRONLY | os.O_CREAT, 0o644)
    try:
        await asyncio.to_thread(fcntl.flock, fd, fcntl.LOCK_EX)
        try:
            yield
        finally:
            await asyncio.to_thread(fcntl.flock, fd, fcntl.LOCK_UN)
    finally:
        os.close(fd)
        inproc.release()

File: ts/src/guard/store.ts
/** Guard registry and tamper-evident ledger persistence. */

import { createHash, randomUUID } from "node:crypto";
import {
  closeSync,
  constants,
  existsSync,
  fsyncSync,
  mkdirSync,
  openSync,
  readFileSync,
  renameSync,
  unlinkSync,
  writeFileSync,
} from "node:fs";
import { homedir } from "node:os";
import { dirname, join } from "node:path";
import { canonicalJson } from "./canonical.js";
import { LedgerCorrupt, ResourceIdMismatch } from "./errors.js";

export type GuardGraph = Record<string, string[]>;
export type EdgePredicates = Record<string, Array<Record<string, unknown>>>;

export type GuardRegistryFields = {
  resource_id: string;
  graph: GuardGraph;
  edge_predicates: EdgePredicates;
  initial: string;
  terminal?: string[];
  stakes?: Record<string, string>;
  checksum?: string;
  graph_version?: number;
  workspace_root?: string | null;
  current_state?: string;
};

export type LedgerEntryFields = {
  ts_ms: number;
  from_state: string;
  to_state: string;
  outcome: string;
  kind: string;
  resolved_by?: string;
  idempotency_key?: string | null;
  payload_digest?: string | null;
  rationale?: string | null;
  verdict?: Record<string, unknown> | null;
  prev_digest?: string;
  entry_digest?: string;
};

export type LedgerEntryCore = Omit<Required<LedgerEntryFields>, "entry_digest">;

// Module-global and read at call time, matching Python's monkeypatchable GUARDS_DIR.
export let GUARDS_DIR = join(homedir(), ".stratum", "guards");

export function setGuardsDir(path: string): void {
  GUARDS_DIR = path;
}

export function resourceHash(resourceId: string): string {
  return createHash("sha256").update(resourceId, "utf8").digest("hex").slice(0, 32);
}

function validateResourceId(resourceId: unknown): asserts resourceId is string {
  if (typeof resourceId !== "string" || resourceId.length === 0) {
    throw new TypeError("resource_id must be a non-empty string");
  }
  if (resourceId.includes("\0")) throw new TypeError("resource_id must not contain NUL");
  if (resourceId === "." || resourceId === "..") throw new TypeError("resource_id cannot be '.' or '..'");
}

export function resourceDir(resourceId: string): string {
  validateResourceId(resourceId);
  return join(GUARDS_DIR, resourceHash(resourceId));
}

const STATE_NAME_RE = /^[A-Za-z0-9_.-]+$/;

export function isValidStateName(name: unknown): name is string {
  return typeof name === "string" && STATE_NAME_RE.test(name);
}

function requiredField<T>(record: Record<string, unknown>, name: string): T {
  if (!Object.hasOwn(record, name)) throw new TypeError(`missing required field ${name}`);
  return record[name] as T;
}

function asRecord(value: unknown): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new TypeError("expected an object");
  }
  return value as Record<string, unknown>;
}

export class GuardRegistry {
  resource_id: string;
  graph: GuardGraph;
  edge_predicates: EdgePredicates;
  initial: string;
  terminal: string[];
  stakes: Record<string, string>;
  checksum: string;
  graph_version: number;
  workspace_root: string | null;
  current_state: string;

  constructor(fields: GuardRegistryFields) {
    this.resource_id = fields.resource_id;
    this.graph = fields.graph;
    this.edge_predicates = fields.edge_predicates;
    this.initial = fields.initial;
    this.terminal = fields.terminal ?? [];
    this.stakes = fields.stakes ?? {};
    this.checksum = fields.checksum ?? "";
    this.graph_version = fields.graph_version ?? 1;
    this.workspace_root = fields.workspace_root ?? null;
    this.current_state = fields.current_state ?? "";
  }

  toDict(): Required<GuardRegistryFields> {
    return {
      resource_id: this.resource_id,
      graph: this.graph,
      edge_predicates: this.edge_predicates,
      initial: this.initial,
      terminal: this.terminal,
      stakes: this.stakes,
      checksum: this.checksum,
      graph_version: this.graph_version,
      workspace_root: this.workspace_root,
      current_state: this.current_state,
    };
  }

  static fromDict(value: unknown): GuardRegistry {
    const record = asRecord(value);
    return new GuardRegistry({
      resource_id: requiredField<string>(record, "resource_id"),
      graph: requiredField<GuardGraph>(record, "graph"),
      edge_predicates: requiredField<EdgePredicates>(record, "edge_predicates"),
      initial: requiredField<string>(record, "initial"),
      ...(Object.hasOwn(record, "terminal") ? { terminal: record.terminal as string[] } : {}),
      ...(Object.hasOwn(record, "stakes") ? { stakes: record.stakes as Record<string, string> } : {}),
      ...(Object.hasOwn(record, "checksum") ? { checksum: record.checksum as string } : {}),
      ...(Object.hasOwn(record, "graph_version") ? { graph_version: record.graph_version as number } : {}),
      ...(Object.hasOwn(record, "workspace_root") ? { workspace_root: record.workspace_root as string | null } : {}),
      ...(Object.hasOwn(record, "current_state") ? { current_state: record.current_state as string } : {}),
    });
  }
}

export class LedgerEntry {
  ts_ms: number;
  from_state: string;
  to_state: string;
  outcome: string;
  kind: string;
  resolved_by: string;
  idempotency_key: string | null;
  payload_digest: string | null;
  rationale: string | null;
  verdict: Record<string, unknown> | null;
  prev_digest: string;
  entry_digest: string;

  constructor(fields: LedgerEntryFields) {
    this.ts_ms = fields.ts_ms;
    this.from_state = fields.from_state;
    this.to_state = fields.to_state;
    this.outcome = fields.outcome;
    this.kind = fields.kind;
    this.resolved_by = fields.resolved_by ?? "agent";
    this.idempotency_key = fields.idempotency_key ?? null;
    this.payload_digest = fields.payload_digest ?? null;
    this.rationale = fields.rationale ?? null;
    this.verdict = fields.verdict ?? null;
    this.prev_digest = fields.prev_digest ?? "";
    this.entry_digest = fields.entry_digest ?? "";
  }

  core(): LedgerEntryCore {
    const { entry_digest: _entryDigest, ...core } = this.toDict();
    return core;
  }

  toDict(): Required<LedgerEntryFields> {
    return {
      ts_ms: this.ts_ms,
      from_state: this.from_state,
      to_state: this.to_state,
      outcome: this.outcome,
      kind: this.kind,
      resolved_by: this.resolved_by,
      idempotency_key: this.idempotency_key,
      payload_digest: this.payload_digest,
      rationale: this.rationale,
      verdict: this.verdict,
      prev_digest: this.prev_digest,
      entry_digest: this.entry_digest,
    };
  }

  static fromDict(value: unknown): LedgerEntry {
    const record = asRecord(value);
    return new LedgerEntry({
      ts_ms: requiredField<number>(record, "ts_ms"),
      from_state: requiredField<string>(record, "from_state"),
      to_state: requiredField<string>(record, "to_state"),
      outcome: requiredField<string>(record, "outcome"),
      kind: requiredField<string>(record, "kind"),
      ...(Object.hasOwn(record, "resolved_by") ? { resolved_by: record.resolved_by as string } : {}),
      ...(Object.hasOwn(record, "idempotency_key") ? { idempotency_key: record.idempotency_key as string | null } : {}),
      ...(Object.hasOwn(record, "payload_digest") ? { payload_digest: record.payload_digest as string | null } : {}),
      ...(Object.hasOwn(record, "rationale") ? { rationale: record.rationale as string | null } : {}),
      ...(Object.hasOwn(record, "verdict") ? { verdict: record.verdict as Record<string, unknown> | null } : {}),
      ...(Object.hasOwn(record, "prev_digest") ? { prev_digest: record.prev_digest as string } : {}),
      ...(Object.hasOwn(record, "entry_digest") ? { entry_digest: record.entry_digest as string } : {}),
    });
  }
}

export function computeEntryDigest(entryCore: LedgerEntryCore | Record<string, unknown>, prevDigest: string): string {
  return createHash("sha256").update(canonicalJson(entryCore) + prevDigest, "utf8").digest("hex");
}

function prettyCanonicalJson(value: unknown): string {
  const compact = canonicalJson(value);
  let output = "";
  let indent = 0;
  let inString = false;
  let escaped = false;

  for (let index = 0; index < compact.length; index += 1) {
    const char = compact[index]!;
    if (inString) {
      output += char;
      if (escaped) escaped = false;
      else if (char === "\\") escaped = true;
      else if (char === '"') inString = false;
      continue;
    }
    if (char === '"') {
      inString = true;
      output += char;
    } else if (char === "{" || char === "[") {
      output += char;
      if (compact[index + 1] !== (char === "{" ? "}" : "]")) {
        indent += 1;
        output += `\n${"  ".repeat(indent)}`;
      }
    } else if (char === "}" || char === "]") {
      if (compact[index - 1] !== (char === "}" ? "{" : "[")) {
        indent -= 1;
        output += `\n${"  ".repeat(indent)}`;
      }
      output += char;
    } else if (char === ",") {
      output += `,\n${"  ".repeat(indent)}`;
    } else if (char === ":") {
      output += ": ";
    } else {
      output += char;
    }
  }
  return output;
}

export function atomicWrite(path: string, content: string): void {
  mkdirSync(dirname(path), { recursive: true });
  const temporaryPath = join(dirname(path), `.${randomUUID()}.tmp`);
  let descriptor: number | undefined;
  try {
    descriptor = openSync(temporaryPath, constants.O_WRONLY | constants.O_CREAT | constants.O_EXCL, 0o600);
    writeFileSync(descriptor, content, "utf8");
    fsyncSync(descriptor);
    closeSync(descriptor);
    descriptor = undefined;
    renameSync(temporaryPath, path);
  } catch (error) {
    if (descriptor !== undefined) closeSync(descriptor);
    try {
      unlinkSync(temporaryPath);
    } catch {
      // The temporary file may not have been created or may already be renamed.
    }
    throw error;
  }
}

export function registryExists(resourceId: string): boolean {
  return existsSync(join(resourceDir(resourceId), "registry.json"));
}

export function persistRegistry(registry: GuardRegistry): void {
  atomicWrite(join(resourceDir(registry.resource_id), "registry.json"), prettyCanonicalJson(registry.toDict()));
}

export function loadRegistryRaw(resourceId: string): GuardRegistry | null {
  const path = join(resourceDir(resourceId), "registry.json");
  if (!existsSync(path)) return null;
  let payload: unknown;
  try {
    payload = JSON.parse(readFileSync(path, "utf8"));
  } catch {
    return null;
  }
  const registry = GuardRegistry.fromDict(payload);
  if (registry.resource_id !== resourceId) {
    throw new ResourceIdMismatch(
      `registry stores resource_id ${JSON.stringify(registry.resource_id)} but ${JSON.stringify(resourceId)} requested`,
    );
  }
  return registry;
}

function ledgerPath(resourceId: string): string {
  return join(resourceDir(resourceId), "ledger.jsonl");
}

const ENTRY_DIGEST_PREFIX_RE = /^\{"entry_digest":"([0-9a-f]{64})",/;

function splitLedgerLines(raw: string): string[] {
  // Python's splitlines() covers these three ledger line endings. Unicode line
  // separators are not expected in canonical JSONL ledger files.
  return raw.split(/\r\n|\r|\n/).filter((line) => line.trim().length > 0);
}

function digestFromRawLine(line: string, prevDigest: string): { expected: string; stored: string } | null {
  const match = ENTRY_DIGEST_PREFIX_RE.exec(line);
  if (match === null) return null;
  const coreCanonical = line.replace(ENTRY_DIGEST_PREFIX_RE, "{");
  const expected = createHash("sha256").update(coreCanonical + prevDigest, "utf8").digest("hex");
  return { expected, stored: match[1]! };
}

export function readLedger(resourceId: string): LedgerEntry[] {
  const path = ledgerPath(resourceId);
  if (!existsSync(path)) return [];
  const lines = splitLedgerLines(readFileSync(path, "utf8"));
  const entries: LedgerEntry[] = [];
  let prev = "";

  for (let index = 0; index < lines.length; index += 1) {
    const isLast = index === lines.length - 1;
    let entry: LedgerEntry;
    try {
      entry = LedgerEntry.fromDict(JSON.parse(lines[index]!));
    } catch {
      if (isLast) break;
      throw new LedgerCorrupt(`ledger line ${index} is not valid JSON (interior)`);
    }
    // JSON.parse collapses Python's 0.0 to 0, so parsed JS numbers cannot
    // reproduce Python's canonical bytes. Ledger lines are already canonical:
    // excising their alphabetically-first entry_digest preserves the exact core.
    const digest = digestFromRawLine(lines[index]!, prev);
    if (digest === null || digest.stored !== digest.expected || entry.prev_digest !== prev) {
      if (isLast) break;
      throw new LedgerCorrupt(`ledger chain broken at line ${index} (interior tampering)`);
    }
    entries.push(entry);
    prev = digest.stored;
  }
  return entries;
}

export function verifyChain(source: string | readonly string[]): boolean {
  // Unlike Python, JavaScript loses the float-vs-int spelling during JSON.parse.
  // Verification must therefore consume raw canonical lines, never parsed entries.
  const lines = typeof source === "string"
    ? splitLedgerLines(existsSync(source) ? readFileSync(source, "utf8") : source)
    : source.filter((line) => line.trim().length > 0);
  let prev = "";
  for (const line of lines) {
    let entry: LedgerEntry;
    try {
      entry = LedgerEntry.fromDict(JSON.parse(line));
    } catch {
      return false;
    }
    const digest = digestFromRawLine(line, prev);
    if (digest === null || entry.prev_digest !== prev || digest.stored !== digest.expected) return false;
    prev = digest.stored;
  }
  return true;
}

export function appendLedger(resourceId: string, entry: LedgerEntry): string {
  const entries = readLedger(resourceId);
  const prev = entries.at(-1)?.entry_digest ?? "";
  entry.prev_digest = prev;
  entry.entry_digest = computeEntryDigest(entry.core(), prev);

  const path = ledgerPath(resourceId);
  mkdirSync(dirname(path), { recursive: true });
  const line = `${canonicalJson(entry.toDict())}\n`;
  const descriptor = openSync(path, constants.O_WRONLY | constants.O_CREAT | constants.O_APPEND, 0o644);
  try {
    writeFileSync(descriptor, line, "utf8");
    fsyncSync(descriptor);
  } finally {
    closeSync(descriptor);
  }
  return entry.entry_digest;
}

export function findByIdempotencyKey(resourceId: string, key: string | null | undefined): LedgerEntry | null {
  if (!key) return null;
  const entries = readLedger(resourceId);
  for (let index = entries.length - 1; index >= 0; index -= 1) {
    const entry = entries[index]!;
    if (entry.idempotency_key === key) return entry;
  }
  return null;
}

export function currentStateFromLedger(entries: LedgerEntry[], initial: string): string {
  for (let index = entries.length - 1; index >= 0; index -= 1) {
    const entry = entries[index]!;
    if (entry.outcome === "applied" || entry.outcome === "deviation") return entry.to_state;
  }
  return initial;
}

export function loadRegistry(resourceId: string): GuardRegistry | null {
  const registry = loadRegistryRaw(resourceId);
  if (registry === null) return null;
  registry.current_state = currentStateFromLedger(readLedger(resourceId), registry.initial);
  return registry;
}

// Slice D uses these through the guard store seam: resourceLock supplies the
// token and assertStillHeld fences the append immediately before it is written.
export {
  LockFenceError,
  LockTimeoutError,
  ProcessIdentityUnverifiableError,
  ResourceLockManager,
  assertStillHeld,
  processIdentity,
  resourceLock,
} from "./lock.js";
export type {
  ProcessIdentity,
  ProcessIdentityProvider,
  ResourceLockHandle,
  ResourceLockManagerOptions,
  ResourceLockOptions,
} from "./lock.js";
